test/issue8/proto.pb.go

```diff
@@ -64,6 +66,11 @@ func (m *Foo) Unmarshal(data []byte) error {
 		}
 		fieldNum := int32(wire >> 3)
 		wireType := int(wire & 0x7)
+
+		if fn := int(fieldNum - 1); fn <= 0 {
```

Comment:
Yes, you understand correctly.

As for unmarshalling: i see no other way honestly. At the time we start parsing - we have no idea what's in the message. As fields come in - we get more info, but still can only answer the question "is this specific field present", but if it's required - that's what we want, and if it's not - we don't care.
And since fields are not always sent in order on the wire - we need to make sense of the whole message before checking.
An alternative approach might be to make 1st pass over the message - to get field types+offsets (handling sub-messages properly) then check and then 2nd pass to copy the data.
But it's much more complex and i would assume - slower for the common case when message is fully correct.

Anyway.
I've updated my implementation with the suggestion given in previous message (simpler code, should be more efficient when parsing as well, since we do more stuff at codegen time). Feel free to use it if you like.
https://github.com/anton-povarov/protobuf/commit/c65b7143b92c5e608cc236533e9dec374d238ec0